sway/lock.c

```diff
@@ -0,0 +1,174 @@
+	output_damage_whole(surf->output);
+}
+
+static void handle_surface_destroy(struct wl_listener *listener, void *data)
+{
+	struct sway_session_lock_surface* surf = wl_container_of(listener, surf, destroy);
+	wl_list_remove(&surf->map.link);
+	wl_list_remove(&surf->destroy.link);
+	output_damage_whole(surf->output);
+	free(surf);
+}
+
+static void handle_new_surface(struct wl_listener *listener, void *data)
+{
+	//struct sway_session_lock_manager *lock_state = server.session_lock->data;
+	struct wlr_session_lock_surface_v1 *lock_surface = data;
+	struct sway_session_lock_surface* surf = calloc(1, sizeof(*surf));
+	if (surf == NULL)
+		return;
+
+	sway_log(SWAY_DEBUG, "new lock layer surface");
+
+	struct sway_output *output = lock_surface->output->data;
+	wlr_session_lock_surface_v1_configure(lock_surface, output->width, output->height);
+
+	surf->surface = lock_surface->surface;
+	surf->output = output;
+	surf->map.notify = handle_surface_map;
+	wl_signal_add(&lock_surface->events.map, &surf->map);
+	surf->destroy.notify = handle_surface_destroy;
+	wl_signal_add(&lock_surface->events.destroy, &surf->destroy);
+}
+
+static void handle_unlock(struct wl_listener *listener, void *data)
+{
+	struct sway_session_lock_manager *lock_state = server.session_lock->data;
+	struct sway_seat *seat;
+
+	sway_log(SWAY_DEBUG, "session unlocked");
+	lock_state->state = SESSIONLOCK_STATE_UNLOCKED;
+	lock_state->lock = NULL;
+	
+	wl_list_remove(&lock_state->new_surface.link);
+	wl_list_remove(&lock_state->unlock.link);
+	wl_list_remove(&lock_state->abandon.link);
+
+	wl_list_for_each(seat, &server.input->seats, link) {
+		seat_set_exclusive_client(seat, NULL);
+		// copied from input_manager -- deduplicate?
+		struct sway_node *previous = seat_get_focus(seat);
+		if (previous) {
+			// Hack to get seat to re-focus the return value of get_focus
+			seat_set_focus(seat, NULL);
+			seat_set_focus(seat, previous);
+		}
+	}
+
+
+	// redraw everything
+	for (int i = 0; i < root->outputs->length; ++i) {
+		struct sway_output *output = root->outputs->items[i];
+		output_damage_whole(output);
+	}
+}
+
+static void handle_abandon(struct wl_listener *listener, void *data)
+{
+	struct sway_session_lock_manager *lock_state = server.session_lock->data;
+	struct sway_seat *seat;
+
+	sway_log(SWAY_INFO, "session lock abandoned");
+	lock_state->state = SESSIONLOCK_STATE_ABANDONED;
+	lock_state->lock = NULL;
+	
+	wl_list_remove(&lock_state->new_surface.link);
+	wl_list_remove(&lock_state->unlock.link);
+	wl_list_remove(&lock_state->abandon.link);
+
+	wl_list_for_each(seat, &server.input->seats, link) {
+		seat_set_exclusive_client(seat, PERMALOCK_CLIENT);
+	}
+
+	// redraw everything
+	for (int i = 0; i < root->outputs->length; ++i) {
+		struct sway_output *output = root->outputs->items[i];
+		output_damage_whole(output);
+	}
+}
+
+static void handle_session_lock(struct wl_listener *listener, void *data)
+{
+	struct sway_session_lock_manager *lock_state = server.session_lock->data;
```

Comment:
I'd prefer to avoid using `data` fields if possible, because of the decreased type-safety. We should be able to use this instead:

```c
struct sway_session_lock_manager *manager = wl_container_of(listener, new_lock, session_locked);
```